### ISSUE TITLE
[CLOSED] Function for fetching gridworld env object(s) by type or name 

### ISSUE THREAD

**Issue by matrx-software** [matrx-software/matrx#15]
<a href="https://github.com/matrxs"><img src="https://avatars1.githubusercontent.com/u/58174304?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [matrxs](https://github.com/matrxs)**
_Tuesday Jan 21, 2020 at 13:03 GMT_
_Originally opened as https://github.com/matrxs/MATRXS/issues/14_

----

ID's of all env objects are now created as: obj_name + random_integer. 
However, this makes it impossible to fetch an object when you don't know the ID of an object. It would be desirable to fetch an object by type (class) or name.

**Comment by matrx-software** [comment]
<a href="https://github.com/olafvisker"><img src="https://avatars3.githubusercontent.com/u/11807080?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [olafvisker](https://github.com/olafvisker)**
_Tuesday Jan 21, 2020 at 13:10 GMT_

----

In grid_world.py change function get_env_object with the following:

```python
    def get_env_object(self, requested_id=None, obj_type=None):
        obj = None

        if requested_id in self.__registered_agents.keys():
            if obj_type is not None:
                if isinstance(self.__registered_agents[requested_id], obj_type):
                    obj = self.__registered_agents[requested_id]
            else:
                obj = self.__registered_agents[requested_id]

        if requested_id in self.__environment_objects.keys():
            if obj_type is not None:
                if isinstance(self.__environment_objects[requested_id], obj_type):
                    obj = self.__environment_objects[requested_id]
            else:
                obj = self.__environment_objects[requested_id]

        if requested_id is None and obj_type is not None:
            for obj_id, env_obj in self.__environment_objects.items():
                if isinstance(env_obj, obj_type): 
                    obj = env_obj
                    break

        return obj
```

**Comment by matrx-software** [comment]
<a href="https://github.com/jwaa"><img src="https://avatars0.githubusercontent.com/u/6059315?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jwaa](https://github.com/jwaa)**
_Wednesday Jan 22, 2020 at 12:46 GMT_

----

> ID's of all env objects are now created as: obj_name + random_integer.
> However, this makes it impossible to fetch an object when you don't know the ID of an object. It would be desirable to fetch an object by type (class) or name.

When did it occur that you wished to fetch an object while you don't know the ID? 

@thaije; please do not use the MATRXS account to post issues, as it makes it more difficult to track the origin and progress of issues.

**Comment by matrx-software** [comment]
<a href="https://github.com/jwaa"><img src="https://avatars0.githubusercontent.com/u/6059315?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jwaa](https://github.com/jwaa)**
_Wednesday Jan 22, 2020 at 12:46 GMT_

----

> 
> 
> In grid_world.py change function get_env_object with the following:
> 
> ```python
>     def get_env_object(self, requested_id=None, obj_type=None):
>         obj = None
> 
>         if requested_id in self.__registered_agents.keys():
>             if obj_type is not None:
>                 if isinstance(self.__registered_agents[requested_id], obj_type):
>                     obj = self.__registered_agents[requested_id]
>             else:
>                 obj = self.__registered_agents[requested_id]
> 
>         if requested_id in self.__environment_objects.keys():
>             if obj_type is not None:
>                 if isinstance(self.__environment_objects[requested_id], obj_type):
>                     obj = self.__environment_objects[requested_id]
>             else:
>                 obj = self.__environment_objects[requested_id]
> 
>         if requested_id is None and obj_type is not None:
>             for obj_id, env_obj in self.__environment_objects.items():
>                 if isinstance(env_obj, obj_type): 
>                     obj = env_obj
>                     break
> 
>         return obj
> ```

@olafvisker; the edit you propose returns the very first object if finds with a certain type. Why only the first and not a list of all those object types?

**Comment by matrx-software** [comment]
<a href="https://github.com/thaije"><img src="https://avatars1.githubusercontent.com/u/2978372?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [thaije](https://github.com/thaije)**
_Friday Jan 24, 2020 at 08:45 GMT_

----

@jwaa Ah yeah sorry about that, I was signed in by default using the MATRXS account by accident. 

For how this function should preferrably work, also see issue #15. Script above was a quick fix for AIMS, in #15 a more permanent solution is proposed which returns a list as you suggest. 

As for the use case: the AIMS project makes use of a Score object, simply named 'score' by the user when creating the scenario. At various locations in the code, e.g. in an action, this Score object needs to be fetched or changed. However, how to fetch this object? The current `get_env_object` requires the `obj_id` to be passed, which is not known. Only that it is an object of class Score with name 'score'.
